Scan operation

gramalingam opened the pull request:
Here is an updated PR for a scan operation based on the ONNX workshop discussions.

gramalingam commented:
A couple of questions where feedback would help:

(1)	The scan axis is basically the sequence axis. Question: do we need to support sequence-axis other than 1 (with batch-axis being 0)? If sequence axis is other than 1, then some form of copying (transposition) will be required in each iteration. Is it useful enough to add?
(2)	If we want only sequence-axis = 1, we can drop the scan_axes attribute. We would then need some other way then to determine the number of scan inputs and number of state variables. We could either add an attribute (number_of_scan_inputs, may be with a default value of 1) or make directions non-optional.
(3)	For variable length inputs, should the specification require the padded values in the output to be zero? Or, can the padded values be “undefined”?

ebarsoum commented on docs/Changelog.md:
```diff
@@ -8224,6 +8224,173 @@ This version of the operator has been available since version 8 of the default O
+<dl>
+<dt><tt>body</tt> : graph (required)</dt>
+<dd>The graph run each iteration. It has N+M inputs: (loop state variables..., scan_input_elts...). It has N+K outputs: (loop state variables..., scan_output_elts...). Each scan_output is created by concatenating the value of the specified scan_output_elt value at the end of each iteration of the loop. It is an error if the dimensions of these values change across loop iterations.</dd>
+<dt><tt>directions</tt> : list of strings</dt>
```
Do we need direction per inputs or one global direction?

gramalingam replied:
I think direction per input is better.

ebarsoum commented on docs/Changelog.md:
```diff
@@ -8224,6 +8224,173 @@ This version of the operator has been available since version 8 of the default O
+  
+      // T.shape[0] denotes the batch-size of T
+      // The batch-size of scan_1, ..., scan_m are all required to be equal
+      batch_size = scan_1.shape[0];
```
Do we need to support [batch, seq,...] and [seq, batch,...] through flag?

gramalingam replied:
I am waiting for other people to provide input … I think this is related to the question I asked up at the top also: do we want to allow flexibility with sequence-axis (and/or batch-axis) or just fix them (batch-axis = 0 and sequence-axis = 1) for now? I am fine either way.

skottmckay commented on docs/Changelog.md:
```diff
@@ -8224,6 +8224,173 @@ This version of the operator has been available since version 8 of the default O
+  The attribute body must be a graph, specifying the computation to be performed in
+  every iteration. It takes as input the current values of the state_variables and
+  the current iterated element of the scan_inputs. It must return the (updated) values
+  of the state_variables and zero or more scan_output_element tensors. The values of the
```
Is the number of output tensors constant across loop iterations, so the body potentially updates none of them rather than returns zero scan_output_element tensors?

gramalingam replied:
Sorry, I don't understand the question. Yes, the number of outputs produced by the body sub-graph is constant across loop iterations. The values of some of these outputs may not vary from loop-iteration to loop-iteration. I didn't understand the point about zero scan_output_element.

skottmckay replied:
If there are zero output elements, what is the purpose of the iterations? Don't they need to produce something, or is it valid to just update the state variables?

gramalingam replied:
We will still return the final-value of state-variables (this will be like the output Y_h of RNN). The scan_output_element tensors are concatenated and returned and that will be like the output Y of RNN.

skottmckay commented on docs/Changelog.md:
```diff
@@ -8224,6 +8224,173 @@ This version of the operator has been available since version 8 of the default O
+<dd>The graph run each iteration. It has N+M inputs: (loop state variables..., scan_input_elts...). It has N+K outputs: (loop state variables..., scan_output_elts...). Each scan_output is created by concatenating the value of the specified scan_output_elt value at the end of each iteration of the loop. It is an error if the dimensions of these values change across loop iterations.</dd>
+<dt><tt>directions</tt> : list of strings</dt>
+<dd>An optional list of M directions. The i-th element of the list specifies the direction (forward or reverse) to be scanned for the i-th scan_input tensor. If omitted, all scan_input tensors will be scanned in the forward direction.</dd>
+<dt><tt>scan_axes</tt> : list of ints (required)</dt>
```
Are there any known use cases where scan_axes is required or could that be added later if necessary?

If required, the value assumes the first dimension of the input is the batch size right, so 0 would be valid and equate to the first axis of each row in the batch?

gramalingam replied:
Yes, I have the same question (your first question) ... and I am hoping somebody will speak up if they think it is needed. Otherwise, let's drop it for now.

dzhulgakov commented on docs/Changelog.md:
```diff
@@ -8224,6 +8224,173 @@ This version of the operator has been available since version 8 of the default O
+
+<dl>
+<dt><tt>body</tt> : graph (required)</dt>
+<dd>The graph run each iteration. It has N+M inputs: (loop state variables..., scan_input_elts...). It has N+K outputs: (loop state variables..., scan_output_elts...). Each scan_output is created by concatenating the value of the specified scan_output_elt value at the end of each iteration of the loop. It is an error if the dimensions of these values change across loop iterations.</dd>
```
does it make sense to also add iteration_number to match Loop's format? It might be useful for doing some masking, but I'm not that sure

ebarsoum replied:
I am not fan of one loop OP with multiple way to terminate.

gramalingam replied:
I believe @dzhulgakov is talking about just providing the iteration number to the loop body, not about adding a trip-count … basically a scan/fold with index. We could do this … but if there is no compelling need right now, we can add further such ops (or extend this) later (just my opinion).

bddppq replied:
IIUC the loop index can be simulate by the loop state variable. This is not ideal, but if we see this pattern come up often than we can add it in a backward compatible way.

gramalingam commented:
I updated the PR as per the discussions above: simplified things by assuming sequence-axis / scan-axis is 1. As Emad says, we can extend this by providing a Boolean flag (for each scan input) in case we want to flip the batch-axis and sequence-axis. We can add this in the future if there is demand.

bddppq commented on onnx/defs/controlflow/defs.cc:
```diff
@@ -227,4 +227,189 @@ ONNX_OPERATOR_SET_SCHEMA(
+		"An optional list of M directions. The i-th element of the list specifies the direction "
+		"(forward or reverse) to be scanned for the i-th scan_input tensor. "
+        "If omitted, all scan_input tensors will be scanned in the forward direction.",
+		AttributeProto::STRINGS,
```
I think it's better to use boolean or enum/int to indicate the direction rather than string.

gramalingam replied:
I agree … but my goal was to be consistent with the current RNN/LSTM/GRU spec. It might be better to change all of these uniformly, if we want to use an enum/int.

bddppq replied:
Here there are only two choices, while for RNN there can be potentially many different activation functions.

gramalingam replied:
I was referring to the "direction: STRING" attribute (Specify if the RNN is forward, reverse, or bidirectional. Must be one of forward (default), reverse, or bidirectional.)

gramalingam replied:
But I can change it to Boolean … but the inconsistency might cause some confusion too.

bddppq commented on onnx/defs/controlflow/defs.cc:
```diff
@@ -227,4 +227,189 @@ ONNX_OPERATOR_SET_SCHEMA(
+		"scan_output_elt value at the end of each iteration of the loop. It is an error"
+		" if the dimensions of these values change across loop iterations.",
+		AttributeProto::GRAPH,
+		true)
```
nit: you don't need to specify `true` for required attributes.

bddppq commented on onnx/defs/controlflow/defs.cc:
```diff
@@ -227,4 +227,189 @@ ONNX_OPERATOR_SET_SCHEMA(
+		" if the dimensions of these values change across loop iterations.",
+		AttributeProto::GRAPH,
+		true)
+	.Attr(
```
I would suggest making this attribute required.

bddppq commented:
@gramalingam ping for regenerating the docs for adding new operators

gramalingam commented:
@bddppq , which docs? I ran "python onnx\defs\gen_doc.py" and added all changed doc files. What am I missing?

bddppq commented:
@gramalingam If you look at the failed CI, it says
```
+git status
HEAD detached at FETCH_HEAD
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)
    
    modified: docs/TestCoverage.md
...
```
Which means TestCoverage.md file is outdated. Please run `python onnx/backend/test/stat_coverage.py` to re-generate it.
@houseroad Maybe add this command to https://github.com/onnx/onnx/blob/master/docs/CONTRIBUTING.md?
